Audiobook Player/Player/SleepTimer.swift

```diff
@@ -0,0 +1,118 @@
+//
+//  SleepTimerManager.swift
+//  Audiobook Player
+//
+//  Created by Florian Pichler on 30.03.18.
+//  Copyright © 2018 Florian Pichler.
+//
+
+import UIKit
+import Foundation
+
+typealias SleepTimerStart = () -> Void
+typealias SleepTimerProgress = (Double) -> Void
+typealias SleepTimerEnd = (_ cancelled: Bool) -> Void
+
+final class SleepTimer {
+    static let shared = SleepTimer()
+
+    let durationFormatter: DateComponentsFormatter = DateComponentsFormatter()
+
+    private var timer: Timer?
+    private var onStart: SleepTimerStart?
+    private var onProgress: SleepTimerProgress?
+    private var onEnd: SleepTimerEnd?
+
+    private let defaultMessage: String = "Pause playback"
+    private let alert: UIAlertController = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+    private var timeLeft: Double = 0.0
+    private let intervals: [Double] = [
+        300.0,
+        600.0,
+        900.0,
+        1800.0,
+        2700.0,
+        3600.0
+    ]
+
+    // MARK: Internals
+
+    private init() {
+        durationFormatter.unitsStyle = .positional
+        durationFormatter.allowedUnits = [ .minute, .second ]
+        durationFormatter.collapsesLargestUnit = true
+
+        reset()
+
+        let formatter = DateComponentsFormatter()
+
+        formatter.unitsStyle = .full
+        formatter.allowedUnits = [ .hour, .minute ]
+
+        alert.addAction(UIAlertAction(title: "Off", style: .default, handler: { _ in
+            self.cancel()
+        }))
+
+        for interval in intervals {
+            let formattedDuration = formatter.string(from: interval as TimeInterval)!
+
+            alert.addAction(UIAlertAction(title: "In \(formattedDuration)", style: .default, handler: { _ in
+                self.sleep(in: interval)
+            }))
+        }
+
+        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
+    }
+
+    private func sleep(in seconds: Double) {
+        onStart?()
+        onProgress?(seconds)
+
+        reset()
+
+        timeLeft = seconds
+        timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(update), userInfo: nil, repeats: true)
+
+        RunLoop.main.add(self.timer!, forMode: RunLoopMode.commonModes)
+    }
+
+    private func reset() {
+        alert.message = defaultMessage
+        
+        guard timer != nil else {
```

Comment:
this shouldn't be necessary, that line after this: `timer?.invalidate()` should only work if timer is != nil, otherwise it won't do anything